democracy/src/tests.rs

```diff
@@ -0,0 +1,684 @@
+		assert_ok!(EncointerDemocracy::submit_proposal(
+			RuntimeOrigin::signed(alice.clone()),
+			proposal_action
+		));
+
+		assert_eq!(EncointerDemocracy::get_electorate(1).unwrap(), 5);
+		assert_eq!(EncointerDemocracy::get_electorate(2).unwrap(), 2);
+	});
+}
+
+#[test]
+fn is_passing_works() {
+	new_test_ext().execute_with(|| {
+		let alice = alice();
+		let cid = register_test_community::<TestRuntime>(None, 10.0, 10.0);
+
+		// electorate is 100
+		let pairs = add_population(100, 0);
+		for p in pairs {
+			EncointerCeremonies::fake_reputation(
+				(cid, 5),
+				&account_id(&p),
+				Reputation::VerifiedLinked,
+			);
+		}
+
+		let proposal_action = ProposalAction::SetInactivityTimeout(8);
+		assert_ok!(EncointerDemocracy::submit_proposal(
+			RuntimeOrigin::signed(alice.clone()),
+			proposal_action
+		));
+
+		// turnout below threshold
+		Tallies::<TestRuntime>::insert(1, Tally { turnout: 1, ayes: 1 });
+		assert_eq!(EncointerDemocracy::is_passing(1).unwrap(), false);
+
+		// low turnout, 60 % approval
+		Tallies::<TestRuntime>::insert(1, Tally { turnout: 10, ayes: 6 });
+		assert_eq!(EncointerDemocracy::is_passing(1).unwrap(), false);
+
+		// low turnout 90 % approval
+		Tallies::<TestRuntime>::insert(1, Tally { turnout: 10, ayes: 9 });
+		assert_eq!(EncointerDemocracy::is_passing(1).unwrap(), true);
+
+		// high turnout, 60 % approval
+		Tallies::<TestRuntime>::insert(1, Tally { turnout: 100, ayes: 60 });
+		assert_eq!(EncointerDemocracy::is_passing(1).unwrap(), true);
+
+		// high turnout 90 % approval
+		Tallies::<TestRuntime>::insert(1, Tally { turnout: 100, ayes: 90 });
+		assert_eq!(EncointerDemocracy::is_passing(1).unwrap(), true);
+	});
+}
+
+#[test]
+fn enactment_updates_proposal_metadata_andenactment_queue() {
+	new_test_ext().execute_with(|| {
+		let cid = create_cid();
+		let alice = alice();
+
+		let proposal_action = ProposalAction::SetInactivityTimeout(8);
+		assert_ok!(EncointerDemocracy::submit_proposal(
+			RuntimeOrigin::signed(alice.clone()),
+			proposal_action
+		));
+
+		let proposal_action2 =
+			ProposalAction::UpdateNominalIncome(cid, NominalIncomeType::from(100i32));
+		assert_ok!(EncointerDemocracy::submit_proposal(
+			RuntimeOrigin::signed(alice.clone()),
+			proposal_action2
+		));
+
+		EnactmentQueue::<TestRuntime>::insert(proposal_action.get_identifier(), 1);
+		EnactmentQueue::<TestRuntime>::insert(proposal_action2.get_identifier(), 2);
+
+		run_to_next_phase();
+		run_to_next_phase();
+		run_to_next_phase();
+
+		assert_eq!(EncointerDemocracy::proposals(1).unwrap().state, ProposalState::Enacted);
+
+		assert_eq!(EncointerDemocracy::proposals(2).unwrap().state, ProposalState::Enacted);
+
+		assert_eq!(EncointerDemocracy::enactment_queue(proposal_action.get_identifier()), None);
+
+		assert_eq!(EncointerDemocracy::enactment_queue(proposal_action2.get_identifier()), None);
+	});
+}
+
+#[test]
+fn enact_update_nominal_income_works() {
```

Comment:
this is the main happy flow test from proposal to enactment. I'd prefer if there was no state faking in this test but real non-unanimous voting action with positive outcome

Reply:
generally i think that it is important to have small **unit** tests that cover one specific aspect of the functionality, the rest can be mocked away for efficiency. in this case we will have enactment unit tests for every possible `ProposalAction` and it would no be efficient IMO to build the whole setup for each of those tests.

nevertheless the input to have a complete happy flow test was really good, as after adding one i discovered that i forgot to update the proposal state on voting which basically makes the whole thing not do anything😂 fixed this now and also added a lazy proposal update extrinsic for the case where everyone has voted and we still need to update the proposal state.